### PR TITLE
only retrieve pubkey once for all validators (partially fixes #4865)

### LINKED ISSUE #4865: Excessive calls to GetPubkey() on external signer
**Tendermint version** 
All, including master

**ABCI app** (name for built-in, URL for self-written if it's publicly available):
gaiad

**Environment**:
- **OS** (e.g. from /etc/os-release): Ubuntu 

**What happened**:
When using an external KMS, GetPubkey() is called a minimum of eight times per block. As the pubkey can reasonably be expected to not change during a given block, this should be cached on the Tendermint side of the connection, rather than pushing packets back and forth over the connection (plus marshalling / unmarshalling overhead - which I appreciate is less now than it previously was). It will also reduce the number of calls to the HSM (where such caching isn't implemented at the KMS) and allow for more concurrent chains to be run on a single HSM before hitting capacity.

**What you expected to happen**:
GetPubkey gets called once, and once only, per block.

**Have you tried the latest version**: yes/no
yes

[edit: once per block, rather than once; thanks @melekes ]

### PR DESCRIPTION
**Pull request by joe-bowman**
## Description

in consensus/state.go, when calulating metrics, retrieve address (ergo, pubkey) once prior to iterating over validatorset to ensure we do not make excessive calls to signer.

Partially closes: #4865

### CODE REVIEW

File: consensus/state.go [hunk 1]
```diff
@@ -1554,39 +1554,41 @@ func (cs *State) recordMetrics(height int64, block *types.Block) {
 		var (
 			commitSize = block.LastCommit.Size()
 			valSetLen  = len(cs.LastValidators.Validators)
+			address types.Address = types.Address{}
```
Comment (marbar3778) [inline consensus/state.go:1557]:
```suggestion
			address types.Address
```

Comment (joe-bowman) [reply consensus/state.go:1557]: what's the standard here; fixes in the PR and squash on merge, or update my branch to have a single squashed commit?

Comment (marbar3778) [reply consensus/state.go]: fixes in PR and then we squash on merge

File: consensus/state.go [hunk 2]
```diff
@@ -1554,39 +1554,41 @@ func (cs *State) recordMetrics(height int64, block *types.Block) {
...
 				commitSize, valSetLen, block.Height, block.LastCommit.Signatures, cs.LastValidators.Validators))
 		}
 
+		if cs.privValidator != nil {
+			pubkey, err := cs.privValidator.GetPubKey()
+			if err != nil {
+				// Metrics won't be updated, but it's not critical.
+				cs.Logger.Error("Error on retrieval of pubkey", "err", err)
+			}
+			address = pubkey.Address()
```
Comment (melekes) [inline consensus/state.go:1570]: this will panic if err != nil above

Comment (joe-bowman) [reply consensus/state.go:1570]: Good catch, ty.

**Comment by tessr** [comment]
Thank you for the issue and the quick fix, @joe-bowman! Would you mind adding a changelog entry so we can credit you in the next release?

**Comment by tessr** [comment]
It would also be cool to have some kind of regression test, but given the way you surfaced this, I'm not sure if that will be straightforward. Thoughts?

**Comment by joe-bowman** [comment]
> Thank you for the issue and the quick fix, @joe-bowman! Would you mind adding a changelog entry so we can credit you in the next release?

done :)

File: CHANGELOG_PENDING.md [hunk 3]
```diff
@@ -84,3 +84,4 @@ Friendly reminder, we have a [bug bounty program](https://hackerone.com/tendermi
 - [types] [\#4764](https://github.com/tendermint/tendermint/pull/4764) Return an error if voting power overflows in `VerifyCommitTrusting` (@melekes)
 - [privval] [\#4812](https://github.com/tendermint/tendermint/pull/4812) Retry `GetPubKey/SignVote/SignProposal` a few times before returning an error (@melekes)
 - [p2p] [\#4847](https://github.com/tendermint/tendermint/pull/4847) Return masked IP (not the actual IP) in addrbook#groupKey (@melekes)
+- [consensus] [\#4895](https://github.com/tendermint/tendermint/pull/4895) Don't excessively query pubkey when collating validator metrics (@joe-bowman)
```
Comment (marbar3778) [inline CHANGELOG_PENDING.md:87]:
```suggestion
- [consensus] [\#4895](https://github.com/tendermint/tendermint/pull/4895) Cache the address of the validator to reduce querying a remote validator (@joe-bowman)
```

?

**Pull request merged**